### ISSUE TITLE
Pre-format citations for CSB

### ISSUE THREAD

**Issue by jmcglone** [mlibrary/heliotrope#1547]
To improve consistency in citations that are provided to users, we'll pre-format them in heliotrope and provide those to cozy-sun-bear to display.

- [x] MLA
- [x] Chicago
- [x] APA
- [ ] ~Create a free-text field that allows for markdown for pre-formatting~

See [CSB#102](https://github.com/mlibrary/cozy-sun-bear/issues/102) for more information on what needs to happen to get info to CSB.

**Comment by mbakeryo** [comment]
We need this for HOB in April 2018.

**Comment by sethaj** [comment]
This ticket/story is about adding a citation field(s) to the Monograph model (and then sending that info to CSB). The idea would be to enter in the citations manuaully (MLA, Chicago, APA). We could add 3 seperate fields, or more likely one field that we'd put a json structure into.

There are a few problems with this:

1) The citation fields are not in the ingest Metadata Template
2) When we ingest 5000 new monographs for HEB, I imagine no one will want to create the 15,000 citations by hand

We could instead automate citations the way Hyrax currently does. There's a lot of code to support this, and MLA, Chicago and APA are specifically included, see https://github.com/samvera/hyrax/blob/master/app/helpers/hyrax/citations_behavior.rb

However, in order to get decent citaitons, we'd need to expose 2 fields in the Metadata Template and Monograph Form. These fields already exist in Hyrax, and actually exist in heliotrope but we either don't use them, or don't require them.

Just for reference, here's a Chicago Monograph citation:

Doniger, Wendy. _Splitting the Difference._ Chicago: University of Chicago Press, 1999.

`Monograph.date_created` exists in heliotrope and is in the Form. However, it's not required. We'd need it to be required, and to make sure it's in the Metadata template. ACLS books have this field in the DLXS header.

`Monograph.based_near` is basically "location" and exists in heliotrope, but we don't have it in the From, so it's always empty `[]`. `.based_near` is a little comlplicated since in Hyrax is actually controlled via the Questioning Authority gem, specifically GEONAMES. We'll need to consider how to handle that (https://github.com/samvera/hyrax/blob/master/app/models/concerns/hyrax/basic_metadata.rb#L29). ALCS books appear to have location (aka: "Prineton, N.J." etc) in the DLXS header.

So I see the current options for citations as:

1) Create them all by hand, store them in the Monograph Model
2) Let Hyrax generate them, don't store them (but we'd need to store date_created and based_near)
3) Because the hyrax generated citations will at some point in time not be good enough for someone eventually, we could add the citation fields to the Monograph, but make them optional. If they are empty, we'll display hyrax generated citations. Or we could do 2, wait for someone to actually have a problem with generated citations, then also do 1.

**Comment by jmcglone** [comment]
Thanks for thinking through all of this and summarizing. **I vote enthusiastically for 2.** I could see where we'd want to display the `date_created` value elsewhere on Fulcrum, like on  publisher catalog pages or as part of the metadata on the monograph catalog page. When we have large collections of books having the publication date will be essential for sorting, filtering, and browsing titles.

As you mention, we are already capturing the location information in DLXS, and this isn't just for ACLS. We have it for pretty much all books on DLXS. So eventually when that stuff migrates, we're going to need a place to put this information. And we definitely don't want to lose it. 

So, in other words, I love the 2 solution and I'm voting for us to continue to capture and store metadata the way we've always been capturing metadata for the things we publish digitally.

**Comment by jmcglone** [comment]
Parking lot discussion on 04/10/18 -- settled on #2 for now but there are a couple outstanding questions.

**Comment by conorom** [comment]
So the Hyrax CitationBehavior code seems quite robust with regard to missing/nonsense fields, so I don't think we need to hinge the display of citations on a whole bunch of metadata checks. Not yet anyway. We can just add the "manual override" field (or fields). Do we need to add those now though?
In the long term, if someone doesn't like the look of a citation, they can either add/correct specific metadata for the Hyrax citations or go all-out and create one (or three) from scratch.
The one field the Hyrax helpers can't do without is `creator`. This isn't great news as we've ignored this field from the get-go. I added a `before_save` to the monograph model which squishes all our primary creator/editor and contributor/editor fields together and uses the resulting array to populate the long-ignored `creator` field. Once that is done you can go straight to the Work Citation page at `/works/999999999/citation` and see what the 3 Hyrax helper citations look like for the work.

I will make a PR for that and then move on to adding any missing, citation-relevant fields to the monograph form.

**Comment by conorom** [comment]
Using the auto-complete feature of the GeoNames field requires setting up a GeoNames account:
https://github.com/samvera/hyrax/wiki/Hyrax-Management-Guide#geonames
In fact, you can't set this field through the UI right now as everything comes back as "No matches found". I guess we could change it to a text box if we don't want this check.
Otherwise we'll set up an account (using developer group email?). Will discuss in a brief parking lot tomorrow.

**Comment by conorom** [comment]
update:
1) I'm removing the `before_save` Monograph model solution as `creator`, being an ActiveTriples::Relation can't retain order. Going to do the same logic as in the `before_save`, essentially, in a `creator` method in the MonographPresenter
2) the GeoNames-based controlled vocab of `based_near` is only enforced on the form. After discussions this morning regarding this enforcement being a possible nuisance (especially with all our content being ingested from outside the UI), I've tested that the enforcement is at the form level only and replaced the form widget with a multi-value text box. This is particularly appropriate given the first note below.

notes:
1) `based_near` is only directly used in Hyrax's `OpenUrlFormatter` (the 4th citation format they have which isn't used anywhere in Hyrax itself), and in `PublicationBehavior.setup_pub_place`, which is not used in the 3 citations mentioned in this ticket either. I guess I'll add `based_near` to our Monograph form anyway
2) `OpenUrlFormatter` directly pulls the `contributor` value and uses it along with `creator`, so if we ever start using this formatter, we'll have to tweak something so that contributors don't appear twice as a result of update 1 above.

**Comment by conorom** [comment]
Usage for this can be seen in this partial, used on the Monograph show page, for when the values are being sent to CSB:
https://github.com/mlibrary/heliotrope/blob/ed1ae488e624c4a62f1a3a54d2bf237a9f252e38/app/views/hyrax/monographs/_csb_hyrax_citations.html.erb
